src/core/internal/arrayop.d

```diff
@@ -0,0 +1,453 @@
+    else version = X86_OR_X64;
+}
+else version (X86_64)
+    version = X86_OR_X64;
+else
+    static assert(0, "unimplemented");
+
+version (GNU) version = GNU_OR_LDC;
+version (LDC) version = GNU_OR_LDC;
+
+T[] arrayOp(T : T[], Args...)(T[] res, Filter!(isType, Args) args) @trusted @nogc pure nothrow
+{
+    enum check = opsSupported!(true, T, Filter!(not!isType, Args)); // must support all scalar ops
+
+    size_t pos;
+    static if (vectorizeable!(T[], Args))
+    {
+        alias vec = .vec!T;
+        alias load = .load!(T, vec.length);
+        alias store = .store!(T, vec.length);
+
+        // Given that there are at most as many scalars broadcast as there are
+        // operations in any `ary[] = ary[] op const op const`, it should always be
+        // worthwhile to choose vector operations.
+        if (res.length >= vec.length)
+        {
+            mixin(initScalarVecs!Args);
+
+            auto n = res.length / vec.length;
+            do
+            {
+                mixin(vectorExp!Args ~ ";");
+                pos += vec.length;
+            }
+            while (--n);
+        }
+    }
+    for (; pos < res.length; ++pos)
+        mixin(scalarExp!Args ~ ";");
+
+    return res;
+}
+
+private:
+
+// SIMD helpers
+
+version (DigitalMars)
+{
+    import core.simd;
+
+    template vec(T)
+    {
+        enum regsz = 16; // SSE2
+        enum N = regsz / T.sizeof;
+        alias vec = __vector(T[N]);
+    }
+
+    void store(T, size_t N)(T* p, in __vector(T[N]) val)
+    {
+        pragma(inline, true);
+        alias vec = __vector(T[N]);
+
+        version (LDC)
+        {
+            storeUnaligned!vec(val, p);
+        }
+        else version (GNU)
+        {
+            static if (is(T == float))
+                __builtin_ia32_storeups(p, val);
+            else static if (is(T == double))
+                __builtin_ia32_storeupd(p, val);
+            else
+                __builtin_ia32_storedqu(cast(char*) p, val);
+        }
+        else version (DigitalMars)
+        {
+            static if (is(T == float))
+                cast(void) __simd_sto(XMM.STOUPS, *cast(vec*) p, val);
+            else static if (is(T == double))
+                cast(void) __simd_sto(XMM.STOUPD, *cast(vec*) p, val);
+            else
+                cast(void) __simd_sto(XMM.STODQU, *cast(vec*) p, val);
+        }
+    }
+
+    const(__vector(T[N])) load(T, size_t N)(in T* p)
+    {
+        import core.simd;
+
+        pragma(inline, true);
+        alias vec = __vector(T[N]);
+
+        version (LDC)
+        {
+            return loadUnaligned!vec(cast(T*) p);
+        }
+        else version (GNU)
+        {
+            static if (is(T == float))
+                return __builtin_ia32_loadups(p);
+            else static if (is(T == double))
+                return __builtin_ia32_loadupd(p);
+            else
+                return __builtin_ia32_loaddqu(cast(const char*) p);
```

Comment:
> That would assume alignment to float4.alignof.

Right.  I should have remembered the segfault in dmd I recently encountered.  Because that is precisely what would happen. ;-)